crypto/s2n_cipher.c

```diff
@@ -23,6 +26,11 @@ int s2n_session_key_alloc(struct s2n_session_key *key)
 {
     eq_check(key->evp_cipher_ctx, NULL);
     notnull_check(key->evp_cipher_ctx = EVP_CIPHER_CTX_new());
+#if defined(S2N_CIPHER_AEAD_API_AVAILABLE)
+    eq_check(key->evp_aead_ctx, NULL);
+    notnull_check(key->evp_aead_ctx = OPENSSL_malloc(sizeof(EVP_AEAD_CTX)));
```

Comment:
In s2n, we try to prevent memory allocation using OPENSSL_malloc and use s2n's memory allocation, is that possible here or is this necessary?

Reply:
@SalusaSecondus 
`key->evp_aead_ctx` is an AWS-LC-specific object that must be allocated using `OPENSSL_malloc`. For the corresponding OpenSSL-specific object `key->evp_cipher_ctx` the allocation is handled internally in the "new" function `EVP_CIPHER_CTX_new`. So, the operations are the same, but in the case of AWS-LC we make the call to `OPENSSL_malloc` directly.

AWS-LC does not have a compatible "new" function. The AEAD API initialise/new function assumes knowledge of which cipher you want to use as well as the key, see AWS-LC AEAD [init function](https://github.com/awslabs/aws-lc/blob/main/third_party/boringssl/include/openssl/aead.h#L243) and [new function](https://github.com/awslabs/aws-lc/blob/main/third_party/boringssl/include/openssl/aead.h#L226). Hence, the abstract cipher implementation in S2N is not compatible with using the AWS-LC AEAD init/new function e.g. the key must be set in a separate function call `s2n_aead_chacha20_poly1305_set_encryption_key`.

I would like to make this change, but it is not possible atm. We can look at adding such an API in the future.

Reply:
@torben-hansen As @andrewhop as approved this I'm willing to accept it. However I still don't understand why ou can't use the standard s2n memory allocation method rather than `OPENSSL_malloc`.